### ISSUE TITLE
Dynamically append available nodes

### ISSUE THREAD

**Issue by netgfx** [chrisjpatty/flume#47]
This is probably a duplicate of #38 but wanted to show some code to be sure. So I have this:

```
const addNode = () => {
        config.addNodeType({
            type: "API_UPDATE",
            label: "API UPDATE",
            description: "Performs an UPDATE HTTP Request",
            inputs: ports => [
                ports.string({ name: "URL", label: "URL" }),
                ports.string({ name: "Name", label: "Name" }),
                ports.string({ name: "Email", label: "Email" }),
                ports.string({ name: "Password", label: "Password" }),
            ],
            outputs: ports => [
                
            ]
        });
    }

    return (

        <div style={{ width: 1024, height: 800 }}>
            <button onClick={saveNodes}>Save Logic</button>
            <button onClick={addNode}>Add Node</button>
            <NodeEditor nodes={nodes} ref={nodeEditor} portTypes={config.portTypes} nodeTypes={config.nodeTypes} />{" "}
        </div>
    );
```
In which I'm trying to dynamically append a node type to the configuration, in order to be available on the right-click menu. Is this something that is doable? It doesn't throw any errors but the list isn't updated.

Thanks!

**Comment by netgfx** [comment]
@chrisjpatty any ideas about this? can it be done?

**Comment by chrisjpatty** [comment]
@netgfx Sorry this fell off my radar. Yes this is possible, all you need is to include the flume config as part of the react state so that updating it will trigger a re-render. Here's an example.

[https://codesandbox.io/s/dynamically-add-flume-types-example-omwgk](https://codesandbox.io/s/dynamically-add-flume-types-example-omwgk)

Even though this is possible, I would caution against it unless absolutely necessary for your use-case. When you create your flume config statically you can guarantee ahead of time that graphs created with that config will fit the schema of your pre-defined types. This means it's safe to import and export nodes created with that config in the root engine and the node editor. If your types are not statically defined, you become responsible for ensuring that everywhere you consume the graphs created with your dynamic configs adhere to a schema that is compatible with your code. For example, if you start with a config which includes a "number" node, but not a "string" node, and some users create graphs with or without the string node, you now need to track which graphs are using the additional node type, and ensure that when you consume the graph that your code for consuming them accounts for the difference in the two schemas. This may be easy enough to do with one or two node types, but if your schema can be quite different between users of your node editor, this could become more difficult to manage. 

So in short, **yes**, this is possible, and there's definitely use-cases where it's the right thing to do, but I'd recommend caution.

**Comment by netgfx** [comment]
Thank you so much for the example and the explanation. So if I save the graph in a neutral place (say database) and all users restore from there, this shouldn't cause any issues right?

**Comment by Abhishek-Kumbhani** [comment]
I think, here the issue is to add a new node from pre-defined node types in FlumeConfig but not the new node type.
